File: java/xrayCalculations/Simulate.java
package fpSimulation;

public class Simulate {
    
    public static void main(String[] args) {
        CSVReader tauReader = new CSVReader("fpSimulation/data/tau.csv");
        double[][] tau = tauReader.toDouble(1,0,0);
        CSVReader muReader = new CSVReader("fpSimulation/data/mu.csv");
        double[][] mu = muReader.toDouble(1,0,0);
        CSVReader RhReader = new CSVReader("fpSimulation/data/rhodium.csv");
        double[][] RhPeaks = RhReader.toDouble(1,1,0);
        CSVReader charReader = new CSVReader("fpSimulation/data/characteristics.csv");
        double[][] characteristics = charReader.toDouble(1,1,0);
        Elements elements = new Elements();
        
        double keVsrc = 9.0;
        double milliamps = 0.250;
        double steradians = 0.015;
        double thetaIn = 78.0;
        double thetaOut = 12.0;
        double contMult = 1.0;
        double charMult = 0.0;
        
        Source source = new Source(keVsrc, milliamps, steradians, thetaIn, thetaOut, contMult, charMult);
        double[][] sourceI = source.calculateIntensity(tau, RhPeaks);
        
        // Path Materials Format : Material#, Z, Ci(wtF), rho(g/cm3), t(cm)
        double[][] beamPath = { {1, 4, 1.000, 1.850, 0.0125},
                                {2, 7, 0.781, 0.001225, 1.000},
                                {2, 8, 0.210, 0.001225, 1.000},
                                {2, 18, 0.009, 0.001225, 1.000},
                                {3, 6, 0.850, 0.940, 0.0004},
                                {3, 1, 0.150, 0.940, 0.0004},
                                {4, 2, 1.000, 0.0001785, 3.000},
                                {5, 6, 0.850, 0.940, 0.0004},
                                {5, 1, 0.150, 0.940, 0.0004} };
                                
        Attenuator beam = new Attenuator();
        double[][] beamI = beam.modifyIntensity(sourceI, beamPath, mu);
        
        double geomFactor = 0.5;
        double[][] sample = { {14, 0.500},{13, 0.250},{8, 0.250} };
        Fluorescence fluor = new Fluorescence(geomFactor);
        double[][] fluorI = fluor.calculateSpectrum(beamI, sample, characteristics, tau, mu);
                                
        double[][] detectorPath = { {1, 6, 0.850, 0.940, 0.0004},
                                    {1, 1, 0.150, 0.940, 0.0004},
                                    {2, 2, 1.000, 0.0001785, 1.000},
                                    {3, 4, 1.000, 1.850, 0.00125} };
        
        int nElements = fluorI.length;
        int nCols = fluorI[0].length;
        for (int i = 0; i < nElements; i++) {
            for (int j = 0; j < nCols; j++) {
                System.out.print(fluorI[i][j] + ", ");
            }
            System.out.print("\n");
        }
        //String[] header = {"keV","I"};
        //CSVWriter sourceWriter = new CSVWriter("fpSimulation/data/source.csv");
        //sourceWriter.writeDoubles(sourceI, header);
        //CSVWriter beamWriter = new CSVWriter("fpSimulation/data/beam.csv");
        //beamWriter.writeDoubles(beamI, header);
        
        return;
    }
}

File: go/src/fpSimulation/simulate.go
package fpSimulation

import (
	"fmt"
)

func Simulate(scen *Scenario, prop *Properties) {
	scen.KeVsrc = 9.0
	scen.MA = 0.250
	scen.Sr = 0.015
	scen.ThetaIn = 78.0
	scen.ThetaOut = 12.0
	scen.ContMult = 1.0
	scen.CharMult = 0.1
	sourceIntensity(scen, prop)

	// Path Materials Format : Material#, Z, Ci(wtF), rho(g/cm3), t(cm)
	beamMaterials := [][]float64{{1.0, 4.0, 1.00, 1.85, 0.0125},
		{2.0, 6.0, 0.85, 0.94, 0.0004},
		{2.0, 1.0, 0.15, 0.94, 0.0004},
		{3.0, 2.0, 1.00, 0.000179, 3.00},
		{4.0, 6.0, 0.85, 0.94, 0.0004},
		{4.0, 1.0, 0.15, 0.94, 0.0004},
		{5.0, 6.0, 0.85, 0.94, 0.0004},
		{5.0, 1.0, 0.15, 0.94, 0.0004}}
	scen.BeamPath = beamMaterials
	attenuateSource(scen, prop)

	// Sample Format : Z, Ci(wtF)
	sample := [][]float64{{14.0, 0.243},
		{13.0, 0.041},
		{26.0, 0.090},
		{20.0, 0.061},
		{12.0, 0.030},
		{11.0, 0.007},
		{22.0, 0.012},
		{19.0, 0.004},
		{8.0, 0.512}}
	scen.Sample = sample
	scen.GeomFactor = 0.5
	calculatePeaks(scen, prop)

	detectorMaterials := [][]float64{{1.0, 6.0, 0.85, 0.94, 0.0004},
		{1.0, 1.0, 0.15, 0.94, 0.0004},
		{2.0, 6.0, 0.85, 0.94, 0.0004},
		{2.0, 1.0, 0.15, 0.94, 0.0004},
		{3.0, 2.0, 1.00, 0.000179, 1.00},
		{4.0, 4.0, 1.00, 1.85, 0.00125},
		{5.0, 14.0, 1.00, 2.16, 0.00005}}
	scen.DetectorPath = detectorMaterials
	attenuatePeaks(scen, prop)

	fwhmSi := 0.100
	fwhmFe := 0.150
	scen.NoiseM = ((fwhmFe / 2.355) - (fwhmSi / 2.355)) / (6.39 - 1.74)
	scen.NoiseB = (fwhmFe / 2.355) - scen.NoiseM*6.39
	mcaSpectrum(scen, prop)

	fmt.Println("Simulation Complete!")
	return
}
